language/bytecode-verifier/src/borrow_graph.rs

```diff
@@ -0,0 +1,230 @@
+            });
+            self.0.insert(to, edge_set.into_iter().collect());
+        }
+    }
+
+    /// removes `nonce` and appropriately concatenates each incoming edge with each outgoing edge of `nonce`
+    pub fn remove_nonce(&mut self, nonce: Nonce) {
+        let nonce_edge_set = self.0.remove(&nonce).unwrap();
+        let mut removed_edges = BTreeMap::new();
+        for (n, es) in &self.0 {
+            removed_edges.insert(
+                n.clone(),
+                es.iter()
+                    .filter(|x| x.to == nonce)
+                    .cloned()
+                    .collect::<Vec<_>>(),
+            );
+        }
+        for (n, es) in &removed_edges {
+            es.iter().for_each(|removed_edge| {
+                let n_edge_set_ref = self.0.get_mut(n).unwrap();
+                n_edge_set_ref.remove(removed_edge);
+                nonce_edge_set.iter().for_each(|nonce_edge| {
+                    if removed_edge.edge_type == EdgeType::Strong {
+                        let mut new_label = vec![];
+                        new_label.append(&mut removed_edge.label.clone());
+                        new_label.append(&mut nonce_edge.label.clone());
+                        let edge = Edge {
+                            edge_type: nonce_edge.edge_type,
+                            label: new_label,
+                            to: nonce_edge.to,
+                        };
+                        n_edge_set_ref.insert(edge);
+                    } else {
+                        let edge = Edge {
+                            edge_type: EdgeType::Weak,
+                            label: removed_edge.label.clone(),
+                            to: nonce_edge.to,
+                        };
+                        n_edge_set_ref.insert(edge);
+                    }
+                });
+            });
+        }
+        checked_postcondition!(!self.0.contains_key(&nonce));
+        checked_postcondition!(self.invariant());
+    }
+
+    /// renames nonces in `self` according to `nonce_map`
+    pub fn rename_nonces(&self, nonce_map: BTreeMap<Nonce, Nonce>) -> Self {
+        let mut new_graph = BTreeMap::new();
+        for (n, es) in &self.0 {
+            new_graph.insert(
+                nonce_map[n],
+                es.iter()
+                    .map(|x| Edge {
+                        edge_type: x.edge_type,
+                        label: x.label.clone(),
+                        to: nonce_map[&x.to],
+                    })
+                    .collect::<BTreeSet<_>>(),
+            );
+        }
+        let new_borrow_graph = BorrowGraph(new_graph);
+        checked_postcondition!(new_borrow_graph.invariant());
+        new_borrow_graph
+    }
+
+    /// checks if `self` is subsumed by `other`
+    pub fn subsumes(&self, other: &Self) -> bool {
+        self.unmatched_edges(other).values().all(|es| es.is_empty())
+    }
+
+    /// joins `other` into `self`
+    pub fn join(&mut self, other: &Self) {
```

Comment:
I don't understand this algorithm. I think we should just blindly union across all of the maps. 

This seems overly complicated

Reply:
This code is implementing the following definition of subsumption from the borrow checker document:

We say G is weakened by H iff H can be obtained from G by performing the following two steps:
Replace an edge p: x -> y, either weak or strong, by a weak edge q: x -> y such that q is a prefix of p.
Add edges to G.

I believe you are suggesting the following alternative definition of subsumption:

We say G is weakened by H iff the set of edges in G is a subset of the set of edges in H.

I believe that both definitions are equally expressive in that they will both allow the same set of programs to be typed.  However, the first one could yield more compact proofs and faster convergence.

Reply:
> I believe that both definitions are equally expressive in that they will both allow the same set of programs to be typed. 

I don't think this is true? But I am a bit confused by the weakening rules. 
```
foo(s: &mut S) {
    let x: &mut u64;
    if (cond()) {
        x = &mut s.f;
    } else {
        x = &mut s.g;
    }
    let other = &mut s.h;
    *other = *x;
    *x = 0;
}
```
I don't see how you weaken and still allow `other = &mut s.h` and its following assignment.  How would this work under weakening?

> However, the first one could yield more compact proofs and faster convergence.

I'm not sure that is true on the convergence part, but I could be wrong. Regardless it seems like an unncessary argument as we plan to try to move to a linear check system. 

The big benefit of the joining system is its way simpler to reason about from a code standpoint 
  - blind unions are far less confusing in my opinion (than having to create new edges)
  - Not having JoinErrors makes join states more principled (and hopefully easier to reason about) and reduces the complexity of the abstract interpreter

Reply:
The current implementation will actually do the union at the join of the then and else branches.  This is because the borrow edges from one side are not subsumed by the other.  The set of unmatched edges will be the entire set.

Reply:
I agree that the convergence, faster or not, becomes irrelevant once we move to a scheme that processes each block exactly once.  But the issue of compact proofs is still relevant.  
Suppose that we use a scheme in which each basic block is annotated with a borrow graph at its beginning.  We process each basic block once starting from the borrow graph at the beginning.  We propagate the borrow graph all the way to the end.  Now we have to check that the information in the resulting borrow graph is no more than in the borrow graph at the beginning of each target.  This check could be either a subset check (which is essentially what you want) or the fancier subsumes check that is in my implementation.  Any choice is sound as long as the information content in the borrow graph at the end of the basic block is stronger than the information content in the borrow graph at the beginning of a target edge. I believe both options being proposed are sound. But the one I am proposing will succeed more often and is potentially more expensive (not sure yet but it may be possible to do something efficient).  The potential advantage of my proposal is that it allows a collection of strong edges (or weak edges for that matter) to be subsumed by a single weak edge.  It is possible to construct examples where this is helpful but ultimately we need either better asymptotic complexity or better performance on examples we are interested in.

Reply:
I see the benefit of subsumes check now. Thanks for bearing with me!